// dva-model-creator/lib/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
__export(require("./actionCreatorFactory"));
class DvaModelBuilder {
    constructor(initState, namespace) {
        this.case = (actionCreator, handler) => {
            this.model.reducers[actionCreator.originType] = (state, action) => handler(state, action.payload);
            return this;
        };
        this.caseWithAction = (actionCreator, handler) => {
            this.model.reducers[actionCreator.originType] = handler;
            return this;
        };
        this.takeEvery = (actionCreator, handler) => {
            return this.setEffects(actionCreator, ({ payload }, effects) => handler(payload, effects));
        };
        this.takeEveryWithAction = (actionCreator, handler) => {
            return this.setEffects(actionCreator, handler);
        };
        this.takeLatest = (actionCreator, handler) => {
            return this.setEffects(actionCreator, [
                ({ payload }, effects) => handler(payload, effects),
                { type: 'takeLatest' },
            ]);
        };
        this.takeLatestWithAction = (actionCreator, handler) => {
            return this.setEffects(actionCreator, [handler, { type: 'takeLatest' }]);
        };
        this.throttle = (actionCreator, handler, ms) => {
            return this.setEffects(actionCreator, [
                ({ payload }, effects) => handler(payload, effects),
                { type: 'throttle', ms },
            ]);
        };
        this.throttleWithAction = (actionCreator, handler, ms) => {
            return this.setEffects(actionCreator, [handler, { type: 'throttle', ms }]);
        };
        this.watcher = (actionCreator, handler) => {
            return this.setEffects(actionCreator, [handler, { type: 'watcher' }]);
        };
        this.build = () => {
            return this.model;
        };
        this.setEffects = (actionCreator, data) => {
            this.model.effects[actionCreator.originType] = data;
            return this;
        };
        this.model = {
            state: initState,
            namespace,
            effects: {},
            reducers: {},
        };
    }
}
exports.DvaModelBuilder = DvaModelBuilder;
